google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java

```diff
@@ -146,24 +176,48 @@ public TransactionRunner readWriteTransaction() {
   public TransactionManager transactionManager() {
     Span span = tracer.spanBuilder(READ_WRITE_TRANSACTION).startSpan();
     try (Scope s = tracer.withSpan(span)) {
-      return pool.getReadWriteSession().transactionManager();
+      return getReadWriteSession().transactionManager();
     } catch (RuntimeException e) {
       TraceUtil.endSpanWithFailure(span, e);
       throw e;
     }
   }
 
   @Override
-  public long executePartitionedUpdate(Statement stmt) {
+  public long executePartitionedUpdate(final Statement stmt) {
     Span span = tracer.spanBuilder(PARTITION_DML_TRANSACTION).startSpan();
     try (Scope s = tracer.withSpan(span)) {
-      return pool.getReadWriteSession().executePartitionedUpdate(stmt);
+      return runWithSessionRetry(
+          SessionMode.READ_WRITE,
+          new Function<Session, Long>() {
+            @Override
+            public Long apply(Session session) {
+              return session.executePartitionedUpdate(stmt);
+            }
+          });
     } catch (RuntimeException e) {
       TraceUtil.endSpanWithFailure(span, e);
       throw e;
     }
   }
 
+  private enum SessionMode {
+    READ,
+    READ_WRITE;
+  }
+
+  private <T> T runWithSessionRetry(SessionMode sessionMode, Function<Session, T> callable) {
+    PooledSession session =
+        sessionMode == SessionMode.READ ? getReadSession() : getReadWriteSession();
+    while (true) {
+      try {
+        return callable.apply(session);
+      } catch (SessionNotFoundException e) {
+        session = pool.replaceReadWriteSession(e, session);
```

Comment:
Following up from my last comment, if the session is READ, is this correct, as replaceReadWriteSession eventually returns 'getReadWriteSession()'?

Reply:
Good point, it wasn't right, even though the method was never called with READ. But that was not correct either, as `executePartitionedUpdate` does not need a read/write session, as it always starts a new transaction (as it needs a special type of transaction).